cpu/sam0_common/include/periph_cpu_common.h

```diff
@@ -680,6 +683,35 @@ typedef struct {
  * destination, are kept in RAM and are read when the channel is enabled and
  * triggered. On the SAML21 platform, these descriptors must reside in the LP
  * SRAM.
+ *
+ * The DMA addresses supplied must point to the **end** of the array to be
+ * transferred. When address increment is enabled this means that the supplied
+ * src or dst argument must point to array + length. When increment is disabled,
```

Comment:
> When address increment is enabled this means that the supplied src or dst argument must point to array + length.
> When increment is disabled, the source or destination address can be used directly. 

Why not just handle this inside the DMA driver? I'd rather avoid creating such pitfalls for API users.

Reply:
I would normally also prefer to handle these kinds of pitfalls inside the code. The two main reason I've made an exception here is because:
- The DMA used normally only by our own peripheral code.
- The DMA path is optimized to be as fast as possible with the least amount of overhead. Adding the extra logic on the DMA side would require something along the lines of:
```
uint8_t *src_addr = incr
                  ? src + (len * (1 << descriptor->BTCTRL.bit.BEATSIZE))
                  : src;
```
With the beat size usually hard coded on the calling peripheral side, the whole retrieval of the beat size and the calculation of the length offset can be simplified or skipped.

So I've tried to compensate by thoroughly documenting this oddity instead of fixing it on the DMA side.

Reply:
So for SPI `(1 << descriptor->BTCTRL.bit.BEATSIZE)` is always 1?
Fine then - I guess I got thrown off a bit by

> When increment is disabled, the source or destination address can be used directly.

But if there is no increment, there is no 'end' to the source address either.

Intuitively this makes more sense: Always pass the end of the DMA transfer buffer. End may be equal to the start of the buffer, then increment is false.

So there are in fact not two different behaviors - it's all the same behavior.

Reply:
>  Fine then - I guess I got thrown off a bit by

I hope to prevent this with the additional docs, let me know if there is still something missing

> Intuitively this makes more sense: Always pass the end of the DMA transfer buffer. End may be equal to the start of the buffer, then increment is false.

So remove the last part of the sentence? ("the source or destination address can be used directly.")

> So there are in fact not two different behaviors - it's all the same behavior.

I'm not sure if this is technically correct, or maybe my explanation is still lacking…
Assuming we have an `uint8_t array[1];` that must be transferred over DMA. wIth `incr = true` and `len = 1`, the source must be passed as `array + 1`. With `incr = false` and `len = 1`, the source must be passed as `array`.

Related question, is it more clear to rename `len` to `num` as function argument, might clarify that it is about the number of transfers, not about the byte length of the data array.

Reply:
> Assuming we have an `uint8_t array[1];` that must be transferred over DMA. wIth `incr = true` and `len = 1`, the source must be passed as `array + 1`. With `incr = false` and `len = 1`, the source must be passed as `array`.

Ah thank you for the clarification - I did indeed get it wrong then.

> is it more clear to rename `len` to `num` as function argument, might clarify that it is about the number of transfers, not about the byte length of the data array.

:+1: 

If I set the beat size to 32 bit and I want to transfer 16 bytes, I would also use `num = 4` and not `len = 16`, right?

Reply:
> If I set the beat size to 32 bit and I want to transfer 16 bytes, I would also use `num = 4` and not `len = 16`, right?

Exactly